src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +230,137 @@ def create(operator, lhs, rhs):
+            # Must be a relational or logical operator. Intrinsic type of
+            # result will be boolean.
+            base_type = BOOLEAN_TYPE
+        else:
+            # We have a numerical operation.
+            if argtypes[0].intrinsic == argtypes[1].intrinsic:
+                if argtypes[0].precision != argtypes[1].precision:
+                    raise NotImplementedError(
+                        f"Cannot determine the type of expression "
+                        f"'{self.debug_string()}' involving arguments of the "
+                        f"same intrinsic type but different precision "
+                        f"('{argtypes[0].precision}' and "
+                        f"'{argtypes[1].precision}').")
+                # Operands are of the same type so that is the type of the
+                # result.
+                base_type = argtypes[0]
+            elif argtypes[0].intrinsic == ScalarType.Intrinsic.REAL:
+                base_type = argtypes[0]
+            elif argtypes[1].intrinsic == ScalarType.Intrinsic.REAL:
+                base_type = argtypes[1]
+
+            # Check that the type of the result is consistent with a numerical
+            # operation.
+            if not base_type or base_type.intrinsic not in (
+                    ScalarType.Intrinsic.INTEGER,
+                    ScalarType.Intrinsic.REAL):
+                for atype in argtypes:
+                    if atype.intrinsic not in (ScalarType.Intrinsic.INTEGER,
+                                               ScalarType.Intrinsic.REAL):
+                        raise TypeError(
+                            f"Invalid argument of type '{atype.intrinsic}' to "
+                            f"numerical operation '{self.operator}' in "
+                            f"'{self.debug_string()}'")
+
+        if all(isinstance(atype, ScalarType) for atype in argtypes):
+            # Both operands are of scalar type.
+            return base_type
+
+        if all(isinstance(atype, ArrayType) for atype in argtypes):
+            # Both operands are of array type.
+            if len(argtypes[0].shape) != len(argtypes[1].shape):
+                raise TypeError(
+                    f"Binary operation '{self.debug_string()}' has operands "
+                    f"of different shape: '{self.children[0].debug_string()}' "
+                    f"has rank {len(argtypes[0].shape)} and "
+                    f"'{self.children[1].debug_string()}' has rank "
+                    f"{len(argtypes[1].shape)}")
+            # In general there is no way we can check that the extents of each
+            # dimension match so we have to assume that they do.
+        shape = (argtypes[0].shape if isinstance(argtypes[0], ArrayType) else
+                 argtypes[1].shape)
+        return ArrayType(base_type, shape=shape)
+
 
 # For automatic API documentation generation
 __all__ = ["Operation", "UnaryOperation", "BinaryOperation"]
```

Comment:
Unless I missed something, the UnaryOperation is not implemented yet and it seems trivial, its always `self.children[0].datatype` regardless of operator. Maybe you could add it in this PR?

Reply:
Oh yes, added.